### DISCUSSION TITLE
Porting MethodBinding / XmlLanguage / Evaluating Binding

### DISCUSSION THREAD

**Discussion by mysteryx93** [AvaloniaUI/Avalonia#7174] [Q&A]
I'm trying to convert MethodBinding from WPF to Avalonia (which should be built-in IMO; handling events in MVVM shouldn't be difficult)

It's mostly converting just fine, but there are a few details that block due to differences in Avalonia.

How do I convert these lines?

[FrameworkElement.Language](https://github.com/mysteryx93/MethodBinding.Avalonia/blob/abb12caa8cdf2c45d3cb3b50e3f0586295bcd316/Src/MethodBinding.Avalonia/MethodBindingExtension.cs#L174) used to determining [Converter locale here](https://github.com/mysteryx93/MethodBinding.Avalonia/blob/abb12caa8cdf2c45d3cb3b50e3f0586295bcd316/Src/MethodBinding.Avalonia/EventArgsExtension.cs#L56)

[TypeConverter(typeof(CultureInfoIetfLanguageTagConverter))](https://github.com/mysteryx93/MethodBinding.Avalonia/blob/abb12caa8cdf2c45d3cb3b50e3f0586295bcd316/Src/MethodBinding.Avalonia/EventArgsExtension.cs#L37)

[Evaluating a binding](https://github.com/mysteryx93/MethodBinding.Avalonia/blob/abb12caa8cdf2c45d3cb3b50e3f0586295bcd316/Src/MethodBinding.Avalonia/PropertyPathHelper.cs#L15) in WPF would use Binding, but in Avalonia it wants an InstancedBinding

Some help converting those few lines would be highly appreciated!

Btw any chance that binding events to methods would be built-in?

**Comment by mysteryx93** [comment]
OK I managed to get the code to compile and tests to run. When I run it in an Avalonia project with a simple binding, however, I get this error

    MediaFinished="{s:MethodBinding MediaFinished}"

> System.InvalidCastException: Unable to cast object of type 'MethodBinding.Avalonia.MethodBindingExtension' to type 'System.EventHandler'.

Does someone know enough of the Avalonia internals to make it work?

[Source code is here](https://github.com/mysteryx93/MethodBinding.Avalonia)

**Reply by maxkatz6** [reply]
I am not a markup extension master, but:
1. You can't return anything but event handler from this markup extension. In some code paths your extension returns itself.
2. Is IProvideValueTarget define in the serviceProvider?

**Comment by mysteryx93** [comment]
The problem is that `provideValueTarget.TargetProperty` returns "MediaFinished" as a string, instead of returning EventInfo or MethodInfo.

Also, the designer throws an error "Property expected" on `{s:MethodBinding MediaFinished}`

**Reply by kekekeks** [reply]
There is no `ldtoken` counterpart for events, so compiler uses strings instead.

**Reply by mysteryx93** [reply]
OK the designer error also happens in WPF and is no problem.

In WPF, provideValueTarget.TargetProperty is MethodInfo, but in Avalonia, it is a string. What can I do from here? Look it up via reflection? I'm not finding a way to get the DataContext.

Any option from here or it's just not supported?

btw what is `ldtoken`

I added a sample project so that others can test and debug it more easily.

**Reply by kekekeks** [reply]
>In WPF, provideValueTarget.TargetProperty is MethodInfo

Compiler is forbidden from using reflection at run time, so it returns just a name. We could potentially add some wrapper class that would provide subscribe/unsubscribe methods like we do with CLR properties - https://github.com/AvaloniaUI/Avalonia/blob/b5ca4734d042aae883630709516b30ae3108fd20/src/Avalonia.Base/Data/Core/ClrPropertyInfo.cs

For now you can use `IProvideValueTarget.TargetObject.GetType().GetEvent(...)`


>I'm not finding a way to get the DataContext.
DataContext isn't usually available yet at this point, so you need to cast `IProvideValueTarget.TargetObject` to `StyledElement` and subscribe to DataContext changes.


>btw what is ldtoken

That's an MSIL opcode that allows one to get `RuntimeTypeHandle`/`RuntimeFieldHandle`/`RuntimeMethodHandle` from a particular type/field/method. e. g. `typeof(String)` is compiled to
```asm
ldtoken [mscorlib]System.String
call class [mscorlib]System.Type [mscorlib]System.Type::GetTypeFromHandle(valuetype [mscorlib]System.RuntimeTypeHandle)
```

**Reply by mysteryx93** [reply]
OOOOHHH that works!!!

At runtime.

    [assembly: XmlnsPrefix("https://github.com/mysteryx93/MethodBinding.Avalonia", "s")]
    [assembly: XmlnsDefinition("https://github.com/mysteryx93/MethodBinding.Avalonia", "MethodBinding.Avalonia")]

It fails if I register like this

        xmlns:s="https://github.com/mysteryx93/MethodBinding.Avalonia"

> Unable to resolve type MethodBinding from namespace https://github.com/mysteryx93/MethodBinding.Avalonia

It works with this

    xmlns:s="clr-namespace:MethodBinding.Avalonia;assembly=MethodBinding.Avalonia"

JetBrains Rider says on TextBlock_Pressed

> Property expected

Any way to disable that warning?

**Comment by mysteryx93** [comment]
It's nearly working; but not quite.

    <media:MediaPlayer IsLoopVisible="False" StopCommandExecuted="{s:MethodBinding {Binding ViewModel, RelativeSource={RelativeSource AncestorType={x:Type Window}}}, MediaUnloaded, {Binding}}">

In this case, it's trying to find StopCommandExecuted on MainView instead of on MediaPlayer!

    var eventInfo = provideValueTarget.TargetProperty as EventInfo ??
        (targetName != null ? provideValueTarget.TargetObject.GetType().GetEvent(targetName) : null);

Any way to access MediaPlayer to search at the right place?

**Reply by kekekeks** [reply]
Make sure that you _immediately_ read both `TargetProperty` and `TargetObject`. IProvideValueTarget implementation is reused across objects, so it will become invalid as soon as the code execution leaves your `ProvideValue` method.

**Comment by mysteryx93** [comment]
It's within ProvideValue, I don't think that's the issue here.
